text/0000-return-position-impl-trait-in-traits.md

```diff
@@ -0,0 +1,508 @@
+}
+
+// Since 'a is named in the bounds, it is captured.
+// `RefIterator` thus becomes:
+
+trait RefIterator for Vec<u32> {
+    type Item<'me>
+    where 
+        Self: 'me;
+
+    type $<'a>: impl Iterator<Item = Self::Item<'a>>
+    where 
+        Self: 'a; // Implied bound from fn
+
+    fn iter<'a>(&'a self) -> Self::$<'a>;
+}
+```
+
+## Validity constraint on impls
+
+Given a trait method where `impl Trait` appears in return position,
+
+```rust
+trait Trait {
+    fn method() -> impl T_0 + ... + T_m;
+}
+```
+
+where `T_0 + ... + T_m` are bounds, for any impl of that trait to be valid, the following conditions must hold:
+
+* The return type named in the corresponding impl method must implement all bounds `T_0 + ... + T_m` specified in the trait.
+* Either the impl method must have `#[refine]`,[^refine] OR
+    * The impl must use `impl Trait` syntax to name the corresponding type, and
+    * The return type in the trait must implement all bounds `I_0 + ... + I_n` specified in the impl return type. (Taken with the first outer bullet point, we can say that the bounds in the trait and the bounds in the impl imply each other.)
+
+[^refine]: `#[refine]` was added in [RFC 3245: Refined trait implementations](https://rust-lang.github.io/rfcs/3245-refined-impls.html). This feature is not yet stable.
+
+Additionally, using `-> impl Trait` notation in an impl is only legal if the trait also uses that notation.
+
+```rust
+trait NewIntoIterator {
+    type Item;
+    fn into_iter(self) -> impl Iterator<Item = Self::Item>;
+}
+
+// OK:
+impl NewIntoIterator for Vec<u32> {
+    type Item = u32;
+    fn into_iter(self) -> impl Iterator<Item = u32> {
+        self.into_iter()
+    }
+}
+
+// OK:
+impl NewIntoIterator for Vec<u32> {
+    type Item = u32;
+    #[refine]
+    fn into_iter(self) -> impl Iterator<Item = u32> + DoubleEndedIterator {
+        self.into_iter()
+    }
+}
+
+// OK:
+impl NewIntoIterator for Vec<u32> {
+    type Item = u32;
+    #[refine]
+    fn into_iter(self) -> std::vec::IntoIter<u32> {
+        self.into_iter()
+    }
+}
+
+// Not OK:
+impl NewIntoIterator for Vec<u32> {
+    type Item = u32;
+    fn into_iter(self) -> std::vec::IntoIter<u32> {
+        self.into_iter()
+    }
+}
+```
+
+### Interaction with `async fn` in trait
+
+This RFC modifies the “Static async fn in traits” RFC so that async fn in traits may be satisfied by implementations that return `impl Future<Output = ...>` as long as the return-position impl trait type matches the async fn's desugared impl trait with the same rules as above.
+
+```rust
+trait Trait {
+  async fn async_fn();
+  
+  async fn async_fn_refined();
+}
+
+impl Trait for MyType {
+  fn async_fn() -> impl Future<Output = ()> + '_ { .. }
+  
+  #[refine]
+  fn async_fn_refined() -> BoxFuture<'_, ()> { .. }
+}
+```
+
+Similarly, the equivalent `-> impl Future` signature in a trait can be satisfied by using `async fn` in an impl of that trait.
+
+## Nested impl traits
```

Comment:
So, this example compiles today:

```rust
#![feature(return_position_impl_trait_in_trait)]

trait Foo {
    fn bar(&self) -> (impl std::fmt::Debug, impl std::fmt::Debug);
}

impl Foo for u32 {
    fn bar(&self) -> (impl std::fmt::Debug, impl std::fmt::Debug) {
        (22, 44)
    }
}
```

this makes sense, given the desugaring described in the RFC, but I didn't find any explicit mention of it (though upon re-reading I realize now it's implicit in sentences like "Each `-> impl Trait` notation appearing in a trait fn return type is effectively desugared to an anonymous associated type. "). We should add an explicit section on this, or perhaps generalize this one to "legal positions for `impl Trait` to appear".

Reply:
I think the intent should be that anything that is legal in a free function should be legal here, so I'll update the section to say that.

Reply:
@nikomatsakis Can you check that the updated text resolves your concerns?